### ISSUE TITLE
Using pre-trained model on custom documents

### ISSUE THREAD

**Issue by Kevmue** [facebookresearch/ELI5#21]
Hey,

I'm trying to use the pre-trained model to answer questions based on some random documents.
As it is mentioned in #11 it should be possible to do so. However, somehow the generate.py s output derives the answer from the original dataset rather than my provided documents.

My current steps:

1. process custom json data with process_data_to_source_target.py
2. apply bpe using subword-nmt
3. run preprocess.py using fairseq
4. run generate.py using fairseq

Details to steps

1. Process data
`python3 ./model_code/process_data_to_source_target.py --input ' +
          input_path + ' --output ' + output_path`

2. Apply bpe using subword-nmt
`subword-nmt apply-bpe -c ' + codes_path + ' < ' +
          output_path+'/train.multitask_source > ' + output_path + '/train.multitask_source_bpe
....
`
3. Run Preprocess.py using fairseq
`python3 ' + preprocess_path + ' --source-lang multitask_source_bpe --target-lang multitask_target_bpe --validpref ' + output_path + '/valid --destdir data-bin/eli5 --srcdict ' +
          source_dict_path + ' --tgtdict ' + target_dict_path`

4. Run Generate.py using fairseq
`python3 ' + generate_path +
          ' data-bin/eli5 --path multitask_checkpoint.pt --gen-subset valid --task translation --nbest 1 --source-lang multitask_source_bpe --target-lang multitask_target_bpe --beam 5 --batch-size 32 --remove-bpe --no-repeat-ngram-size 3 --max-len-b 500 --min-len 200 --max-source-positions 4096 --max-target-positions 4096 --skip-invalid-size-inputs-valid-test --model-overrides "{\'max_source_positions\':4096, \'max_target_positions\':4096}" --cpu --results-path data-bin/eli5/results`

Thanks for your answer

**Comment by huihuifan** [comment]
hello, did you replace the path argument in your generate.py call with your custom documents? you can also try using interactive.py instead of generate.py and cat in your custom documents

**Comment by Kevmue** [comment]
Thanks for your quick answer.

My fully compiled command for using generate.py is the following:
`python3 ./fairseq/generate.py data-bin/eli5 --path multitask_checkpoint.pt --gen-subset valid --task translation --nbest 1 --source-lang multitask_source_bpe --target-lang multitask_target_bpe --beam 5 --batch-size 32 --remove-bpe --no-repeat-ngram-size 3 --max-len-b 500 --min-len 200 --max-source-positions 4096 --max-target-positions 4096 --skip-invalid-size-inputs-valid-test --model-overrides "{'max_source_positions':4096, 'max_target_positions':4096}" --cpu --results-path data-bin/eli5/results`

My preprocessed data using preprocess.py is stored in the data-bin/eli5 directory.

The --path argument points to the pre-trained model.

I thought the custom documents are incorporated into the generated ...source_bpe.bin file.

Do you need further information to point me to the right direction?

Thanks a lot

**Comment by Kevmue** [comment]
It seems like the path argument is used for specifying the path to the pre-trained model. Is there another way to provide my custom support documents to the pre-trained model?

**Comment by huihuifan** [comment]
One problem I can see in your argument is this:
`--gen-subset valid`

this means in the data-bin, the model is looking for files like valid. multitask_source_bpe and valid.multitask_target_bpe in a binarized form. This looks suspicious to me, depending on how you preprocessed your data. 

Can you try using interactive.py instead of generate.py? In fairseq, interactive.py allows users to interactively enter text.

Here's what I did:

1. git cloned fairseq
2. built fairseq: pip install --editable .
3. built the dictionary, etc for eli5
4. python interactive.py /angelafan/eli5_open_source/model_code --path /angelafan/eli5_open_source/multitask_checkpoint.pt --source-lang multitask_source_bpe --target-lang multitask_target_bpe --no-repeat-ngram-size 3 

where the directory model_code contains dict.multitask_source_bpe.txt and dict.multitask_target_bpe.txt. 

This should work with generate.py as well, but likely the data you binarized is not correctly named or in the wrong place, etc.

**Comment by huihuifan** [comment]
for completeness, you could do

cat my_custom_questions_and_supportdocs_in_correctformat.txt | python interactive.py ... 
and set arguments like --batch-size and --buffer-size so fairseq batches the generation.

but you can also use this to debug why your generate.py is not working - it looks like a path problem to me.

**Comment by Kevmue** [comment]
@huihuifan thank you for your response

I've tested to generate answers using the interactive.py script. I've created a sample question with some related documents. However, the answer seems to be derived from a ELI5 / Reddit dataset and does not seem to be based on the provided support documents. Is this expected (see attached screenshot)?

I'd expect that the answer includes elements from the support document.

I've created a screenshot of the interactive.py session:

![image](https://user-images.githubusercontent.com/17836114/75110679-4fc63f00-5631-11ea-81de-8b6ecc8f0f02.png)

**Comment by Kevmue** [comment]
Thanks a lot for your support  @huihuifan 

Is this an expected behaviour when using the pre-trained-model? 
I thought that the derived answers are based on the support documents and not completely based on the trained data.

Or is there another issue. Since I was using the interactive.py script a path issue can be excluded, right? Is there an issue with the format of the input data (screenshot)?
